### ISSUE TITLE
How should file tangling details be specified?

### ISSUE THREAD

**Issue by Luis-Henriquez-Perez** [aldrichtr/tangld#12]
Right now I'm thinking of an list of lists.

`(regexp-or-fn list-of-paths-to-tangle-to whether-to-tangle-or-not)`

There should also be a variable for the directory to link to (most of the time it will be home I'm guessing)

Once the files in the dotfiles directory should then be linked to the home directory (or whatever the user specifies).

There should be an option to write this information to a file somehow so it can be version controlled and read.

**Comment by Luis-Henriquez-Perez** [comment]
I decided just to have a user option that takes a function. The function will take a file path and decide what to do with it. It could do nothing, or it could tangle it to the dotfiles dir, or it could just link it to the dotfiles dir. Whatever the user wants.

**Comment by aldrichtr** [comment]
tangld-build needs a list of one or more org files. For each file in the list, tangld-build runs org-bable-tangle on it, and then moves to the next file.  At the most basic level, the src blocks in the file define the options and the target that are given to the org-babel-tangle-file function.  If desired, the user can create other src-blocks in other files that can add additional functions and/or data to the current file's src-blocks.  This is done by loading the library-of-babel which is documented on [the org site](https://orgmode.org/manual/Library-of-Babel.html#Library-of-Babel).

By default, we shouldn't tangle a file that hasn't changed since the last tangle.  This should help reduce "build time" and will also keep file timestamps consistent with changes for the most part.  Also, we want to give the user the ability to include and exclude any file they want from the list that tangld-build uses.  

That said, tangld-build needs the following information:
- A source-directory :: directory where the source-files are (recursively descending by default)
- A glob pattern :: any file in the directory that matches this pattern will be added to the list ('\\.org' by default)
- A file-exclude-regex :: any file that matches this regex would get "skipped" by tangld-build.
- A file-include-regex :: any file that matches this regex would be added back in (like an exclude from the exclude)

to put it all together:
if i have my org files in ~/.tangld/src, with the glob pattern as "\\.org", the exclude-regex as "^-" and include-regex as "^-default", then:
- ~/.tangld/src/tmux.org : build if changed since last time
- ~/.tangld/src/-tmux-powerline-testing.org : skipped, matches "^-"
- ~/.tangld/src/-default-tmux-terminal.org : build if changed since last time (matches "^-default"

Have a look at https://github.com/aldrichtr/tangld/blob/2d3f31128167c71e17ff0098e971f3ef0059a4e3/tangld-build.el

**Comment by Luis-Henriquez-Perez** [comment]
> I like that idea. Originally, I was thinking that within the org file

Ok. Yea good point.

Also, we need to consider non-org files. It is not unlikely that the user will have at least a few non-org files which (because they're not org files) cannot be tangled. So we need to give the user a way to account for those.

**Comment by Luis-Henriquez-Perez** [comment]
> these values can also have a relative directory, or a path below a root-directory if we use ob-extended-tangle. However, it may also be useful to be able to "redirect" that target to a test or staging area, is that what you were thinking?

Yes, I have been thinking about this more. I'd like users to have the ability to "redirect" these files wherever they want. They'd be able to specify a function which takes a target such as ~/.tmux.conf and return the path it should be tangled to.

**Comment by aldrichtr** [comment]
I like that idea. So we need to parse an org file, gather up the tangle options embedded in it and then do some manipulation?
```
defun org-babel--get-vars (params)
  "Return the babel variable assignments in PARAMS.

PARAMS is a quasi-alist of header args, which may contain
multiple entries for the key `:var'.  This function returns a
list of the cdr of all the `:var' entries."
```

Looks like we may be able to use this function out of ob-core ?
